[WIP] Add new proc_macro "node_test"

adrianbenavides opened the pull request:
This PR is motivated by the need to refactor the transports crates (abstract the generic functionality, etc). I think the priority would be to have some tests in place before continuing this task.

In a nutshell, this PR adds a macro parser that is used both by the `node` and `node_test` attributes, following the `tokio`'s macros. Since it's the first macro I've ever done, I've put these changes in a new crate to avoid overwriting the current `node` macro. Ideally, the changes presented could be merged into the `ockam_node_attribute` crate and delete this one.

Here are some details about the attributes:

`node`:
    - refactored so that the macro modifies the input function, instead of creating a new function from scratch (the `trampoline` function is not needed now)
    - adds a line before exiting the `executor`'s block to stop the `context` when the input function is not using a mutable instance (in this case the context can't be stopped from the input function)
    - includes the same tests as found in the `ockam` crate to check that this refactor doesn't break anything (hopefully)

`node_test`:
    - similar to the `node` attribute
    - includes a `timeout` argument to make the test fail after a given time. You can check how it works in the tcp test
    - note that the `node_test` has the same limitations as described in https://github.com/ockam-network/ockam/issues/1576. That is, if an assertion fails, the test will hang as the node keeps running.
    - the `trybuild` tests for "pass" cases check a compile error (the main function doesn't exist). I don't know if there is any other way to test macro attributes that are meant to be used in tests

## Questions/doubts

- Is it OK to add the "auto-stop context" line? I don't see why not, but maybe you have some special use cases in mind.
- The more I look at these changes, the less confident I am about them :laughing: Do you think the `trybuild` tests are enough? Is there any tricky use case of the `node` attribute you can think of?

## Related issues

https://github.com/ockam-network/ockam/issues/1576
https://github.com/ockam-network/ockam/pull/1768

SanjoDeundiak commented:
Hey @adrianbenavides . Thank you for submitting the PR ❤️. It'll take us some time to fully review it.
Meanwhile,
> Is it OK to add the "auto-stop context" line? I don't see why not, but maybe you have some special use cases in mind.

No, there are cases (most of the times, actually) you want your node to run forever.

adrianbenavides commented:
> Hey @adrianbenavides . Thank you for submitting the PR heart. It'll take us some time to fully review it. Meanwhile,
> 
> > Is it OK to add the "auto-stop context" line? I don't see why not, but maybe you have some special use cases in mind.
> 
> No, there are cases (most of the times, actually) you want your node to run forever.

OK got it, but would it make sense for tests only? In case the test function doesn't use a `mut Context`.

adrianbenavides commented:
Btw @SanjoDeundiak, this PR would contain fewer changes if I merged them into the `ockam_node_attribute` subcrate and it would probably be easier to review. Let me know if you want me to move them there!

SanjoDeundiak reviewed (changes requested):
Reviewed most of the PR. Looks very cool!
Could you please look at why tests are failing?

SanjoDeundiak commented on implementations/rust/ockam/ockam_node_test_attribute/src/parser.rs:
```diff
@@ -0,0 +1,128 @@
+
+use proc_macro2::{Ident, Span, TokenStream as TokenStream2};
+use quote::{quote, quote_spanned, ToTokens};
+use syn::token;
```
I think if this unused import is removed, tests become green.

adrianbenavides replied:
Yeah, I missed that one, thanks!

SanjoDeundiak commented on implementations/rust/ockam/ockam_node_test_attribute/tests/trybuild.rs:
```diff
@@ -0,0 +1,7 @@
+    let t = trybuild::TestCases::new();
+    t.pass("tests/node/pass*.rs");
+    t.compile_fail("tests/node/fail*.rs");
+    t.compile_fail("tests/node_test/*.rs");
```
Maybe worth adding a comment why tests that start with "pass_" goes under compile_fail all here. Wasn't obvious for me

adrianbenavides replied:
100% agree to add it here too (I initially added this to the PR description)

SanjoDeundiak commented:
Very nice test coverage. I'll go through every case tomorrow

SanjoDeundiak commented on implementations/rust/ockam/ockam_node_test_attribute/tests/node/pass_when_used_on_any_fn.rs:
```diff
@@ -0,0 +1,4 @@
+#[ockam_node_test_attribute::node]
+async fn foo(mut c: ockam::Context) {
```
I think because macro will expand into main() function, it makes sense to me that function used with this macro must be called main. E.g. this won't compile, which is misleading
```rs
#[ockam_node_test_attribute::node]
async fn foo(ctx: &Context) {
    ctx.stop().await.unwrap();
}

fn main() {}
```rs
This doesn't compile, which is understandable
```
#[ockam_node_test_attribute::node]
async fn main(ctx: &Context) {
    ctx.stop().await.unwrap();
}

fn main() {}
```

adrianbenavides replied:
"must be called main" as we should force it in the macro? or just change the naming in the tests?

adrianbenavides replied:
OK, I got you. Makes total sense.

SanjoDeundiak commented on implementations/rust/ockam/ockam_node_test_attribute/tests/node_test/pass_with_non_mutable_context.rs:
```diff
@@ -0,0 +1,4 @@
+#[ockam_node_test_attribute::node_test]
+async fn my_test(ctx: Context) {
```
What's the difference between this and ```pass_with_owned_context.rs```?

adrianbenavides commented:
Thanks @SanjoDeundiak for the thorough review! I'll go one by one and solve them as soon as I can.

One general question. As I opened this as a WIP, I didn't add the gradle jobs to execute the new crate's tests. I'll try to do this as well and reach out if I get lost.

adrianbenavides commented on implementations/rust/ockam/ockam_node_test_attribute/src/parser.rs:
```diff
@@ -0,0 +1,128 @@
+    };
+    let ctx_ident = &ctx_pat.ident;
+    // If the input function doesn't use a mutable Context, it won't be able to stop it.
+    let ctx_stop_stmt = if ctx_pat.mutability.is_none() {
```
This was just to make it easier to write the tests somehow, but I agree it moves the complexity to the macro, which is bad. Also, without this, the tests will be more explicit regarding how the context is used, so let's remove this part.

adrianbenavides replied:
Should we then force a mutable borrow in the `node_test` attribute? Currently, there is no limitation in this regard.

adrianbenavides replied:
The thing for test functions is that for happy-path tests this becomes easier to handle because you can just stop the context at the end of the test. But if the test panics, how do we stop the context? This should be done by the macro, otherwise we are forcing the user to wrap the test in another function so they can stop the context if it failed... which is weird. Makes sense?

SanjoDeundiak replied:
But it's important that test is not hanging because of panic/failure. Timeout should deal with that just fine

adrianbenavides commented on implementations/rust/ockam/ockam_node_test_attribute/src/parser.rs:
```diff
@@ -0,0 +1,128 @@
+                    let (mut #ctx_ident, mut executor) = ockam_node::start_node();
+                    executor
+                        .execute(async move {
+                            let is_ok = timeout(Duration::from_millis(#timeout as u64), async #body).await.is_ok();
```
Not sure I'm understanding the `tokio::select` you are describing, but it reminds me of this pattern: https://docs.rs/tokio/0.2.13/tokio/macro.select.html#avoid-racy-if-preconditions (the second example without the race condition). If that's what you mean, isn't that equivalent to use `tokio::timeout`?

SanjoDeundiak replied:
What if use this
```rs
use core::time::Duration;
use tokio::time::timeout;

let (mut #ctx_ident, mut executor) = ockam_node::start_node();
let res = executor
    .execute(async move {
        let is_ok = timeout(Duration::from_millis(#timeout as u64), async #body).await.is_ok();
        if is_ok {
            Ok(())
        }
        else {
            Err(ockam_core::Error::new(0, "")) // FIXME: Proper error should be used
        }
    })
    .expect("Executor failed");
assert!(res.is_ok())
```

This way you don't need extra channel to notify about test result, you can just return result to outer synchronous scope. The only thing, it's requires some update to executor::execute() implementation to forward that result, but I already tried implementing that, and it works

adrianbenavides commented on implementations/rust/ockam/ockam_transport_tcp/tests/send_receive.rs:
```diff
@@ -43,13 +24,12 @@ async fn run_test(ctx: &Context) -> Result<()> {
+        let reply = ctx.receive::<String>().await.unwrap();
         assert_eq!(reply, msg, "Should receive the same message");
-        ctx.stop().await?;
+        ctx.stop().await.unwrap();
```
Probably not needed here, but I have a question related to the comment above: https://github.com/ockam-network/ockam/pull/1904#discussion_r722989794

How bad it is to stop the context twice?

SanjoDeundiak commented on implementations/rust/ockam/ockam_transport_tcp/src/lib.rs:
```diff
@@ -11,7 +11,6 @@
 
 #![deny(
     missing_docs,
-    dead_code,
```
Interesting. Latest develop works fine for me and ```dead_code``` check is enabled...

adrianbenavides commented:
Hi @SanjoDeundiak, I finally had some time this morning to keep working on this! As a summary:

* I've done some refactors to split the logic between the test and non-test macros. It should be easier to follow the code now, as most of the complexity is on the node_test macro
* I've added more checks to both macros, you can see them in the tests
* I've added the basic files that every subcrate must have

What I haven't done:
* Add the new subcrate to gradle to run the tests. I took a look at the `rust/build.gradle` but couldn't see if I have to do something to add it, or it's done automatically
